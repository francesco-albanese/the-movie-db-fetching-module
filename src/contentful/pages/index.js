import { 
  contentTypes,
  contentfulClient, 
  urls 
} from '#contentful/contentful.const'
import { transformPageResponse } from '#transformation'
import { parseStringifiedData } from '#utils'

export const getPages = (locale = '*') => {
  return contentfulClient.get(urls.entries, {
    params: {
      content_type: contentTypes.page,
      include: 3,
      locale,
      select: 'sys.id,fields.name,fields.reference,fields.sections'
    },
    transformResponse: [
      data => parseStringifiedData(data),
      data => transformPageResponse(data)
    ]
  })
    .then(data => data)
}