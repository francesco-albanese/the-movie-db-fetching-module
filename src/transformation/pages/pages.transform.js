import { isEmpty } from 'lodash-es'
import { getFirstObjectKey, getSections } from '#utils'

export const getPaths = (path = {}) => {
  const pathList = Object.entries(path)

  return pathList.reduce((paths, [ locale, path ]) => {
    paths[ locale.substring(0, 2) ] = path
    return paths
  }, {})
}

export const transformPageResponse = data => {
  if (isEmpty(data)) {
    return []
  }

  const { items, includes } = data

  return items.map(page => {

    const { 
      name,
      path,
      reference,
      sections 
    } = page.fields
    
    const parsedSections = !isEmpty(sections) 
      ? getSections(page, includes.Entry)
      : null

    const paths = getPaths(path)

    const transformedPage = {
      name: name[ getFirstObjectKey(name) ],
      paths,
      reference: reference[Â getFirstObjectKey(reference) ],
      sections: parsedSections
    }
    
    return transformedPage
  })

}