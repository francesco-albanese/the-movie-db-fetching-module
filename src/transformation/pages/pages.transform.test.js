import { transformPageResponse } from './pages.transform'
import * as Fixtures from './__fixtures__/pages.transform.fixtures'

describe('transformPageResponse', () => {
  it('returns an empty array if data is missing', () => {
    expect(transformPageResponse()).toEqual([])
  })

  it('returns transformed pages', () => {
    const result = transformPageResponse(Fixtures.data)
    expect(result).toMatchSnapshot()
  })
})