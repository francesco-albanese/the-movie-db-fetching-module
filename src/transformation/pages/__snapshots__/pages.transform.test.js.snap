// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformPageResponse returns transformed pages 1`] = `
Array [
  Object {
    "name": "FavouritesPage",
    "paths": Object {
      "en": "/en/favourites",
      "it": "/it/favoriti",
    },
    "reference": "favourites-page",
    "sections": Object {
      "Favourites": Object {
        "MyFavourites": Object {
          "lineOne": Object {
            "en": "My Favourites",
            "it": "I miei preferiti",
          },
          "name": Object {
            "en": "MyFavourites",
          },
        },
      },
    },
  },
  Object {
    "name": "HomePage",
    "paths": Object {
      "en": "/en",
      "it": "/it",
    },
    "reference": "home-page",
    "sections": Object {
      "SearchMovies": Object {
        "SearchMoviesText": Object {
          "lineOne": Object {
            "en": "Search Movies",
            "it": "Cerca Film",
          },
          "name": Object {
            "en": "SearchMoviesText",
          },
        },
      },
    },
  },
  Object {
    "name": "MovieDetailsPage",
    "paths": Object {
      "en": "/en/movie/:movieid",
      "it": "/it/film/:movieid",
    },
    "reference": "movie-details-page",
    "sections": Object {
      "MovieDetails": Object {
        "Cast&Crew": Object {
          "lineOne": Object {
            "en": "Cast and crew",
            "it": "Squadra produzione",
          },
          "lineTwo": Object {
            "en": "Actors",
            "it": "Attori",
          },
          "name": Object {
            "en": "Cast&Crew",
          },
        },
        "DirectorsProducers": Object {
          "lineOne": Object {
            "en": "Director",
            "it": "Regista",
          },
          "lineTwo": Object {
            "en": "Producers",
            "it": "Produttori",
          },
          "name": Object {
            "en": "DirectorsProducers",
          },
        },
        "Genre": Object {
          "lineOne": Object {
            "en": "Genre",
            "it": "Genere",
          },
          "lineTwo": Object {
            "en": "Released",
            "it": "Lancio",
          },
          "name": Object {
            "en": "Genre",
          },
        },
        "Overview": Object {
          "lineOne": Object {
            "en": "Overview",
            "it": "Breve Descrizione",
          },
          "lineTwo": Object {
            "en": "Information",
            "it": "Informazioni",
          },
          "name": Object {
            "en": "Overview",
          },
        },
        "Runtime": Object {
          "lineOne": Object {
            "en": "Runtime",
            "it": "Durata",
          },
          "lineTwo": Object {
            "en": "Rating",
            "it": "Voto",
          },
          "name": Object {
            "en": "Runtime",
          },
        },
        "Title": Object {
          "lineOne": Object {
            "en": "Title",
            "it": "Titolo",
          },
          "lineTwo": Object {
            "en": "Tagline",
            "it": "Citazione",
          },
          "name": Object {
            "en": "Title",
          },
        },
      },
    },
  },
]
`;
